### ISSUE TITLE
Error when unmounting and ad is not shown

### ISSUE THREAD

**Issue by TomiTakussaari** [jaanauati/react-dfp#124]
First time using Google Ads and DFP, so not exactly sure where the problem is.
But when changing route in our app, I get this kind of error:

```
Uncaught TypeError: Cannot read property 'gptSlot' of undefined
    at manager.js:312
    at Array.map (<anonymous>)
    at EventEmitter.destroyGPTSlots (manager.js:311)
    at EventEmitter.unregisterSlot (manager.js:361)
    at AdSlot.unregisterSlot (adslot.js:142)
```
It looks like this happens because manager. registeredSlots has entries that do not (yet?) have ```gptSlot```. 

I presume that gptSlot is somehow given in async way, which leads to this error, but I could be wrong..

**Comment by jaanauati** [comment]
Hi @TomiTakussaari, couple of questions:
 -  what version of react-dfp are u using (according the line numbers does not seem to be the latest one)?
 - could u please share an example of your code?

Thanks.

**Comment by TomiTakussaari** [comment]
Using react-dfp 0.10.2:
```
Uncaught TypeError: Cannot read property 'gptSlot' of undefined
    at manager.js:317
    at Array.map (<anonymous>)
    at EventEmitter.destroyGPTSlots (manager.js:316)
    at EventEmitter.unregisterSlot (manager.js:380)
    at AdSlot.unregisterSlot (adslot.js:142)
    at AdSlot.componentWillUnmount (adslot.js:77)
    at callComponentWillUnmountWithTimer (react-dom.development.js:16699)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at safelyCallComponentWillUnmount (react-dom.development.js:16706)
    at commitUnmount (react-dom.development.js:17083)

```

Code:

```
<AdBanner sizes={[[300, 250]]} slotId="my-slotid" adUnit="foobar"/>
```
We have 0-N AdBanner components per page (route), it renders AdSlot if ads are allowed for customer.

```
class AdBanner extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      error: false,
    };
  }
  public componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    logger.error({ error, errorInfo }, 'Error rendering add');
    this.setState({ error: true });
  }

  public render(): JSX.Element {
    if (this.state.error) {
      return null;
    }
    return (
      <AdsContext.Consumer>
        {({ adsAllowed }) => {
          return (adsAllowed ? <AdSlot {...this.props} /> : null);
        }}
      </AdsContext.Consumer>
    );
  }
}

export default AdBanner;

```

We are using Next.js, and have DFPSlotsProvider in our _app.tsx.
```
          <DFPSlotsProvider
            dfpNetworkId={this.props.dfpNetworkId}
            collapseEmptyDivs={true}
            autoLoad={!this.props.disableAutoLoad}
          >
```

AdsContext is React context that knows if customer has not allowed ads for our site

**Comment by TomiTakussaari** [comment]
Oh, and thanks for quick response :)

**Comment by TomiTakussaari** [comment]
Hmm, I believe this was fixed when I removed slotId property from AdSlot.

**Comment by jaanauati** [comment]
makes sense, my guess is that there might be a buggy behavior in react-dfp when reusing slotIds, if this is not a requirement for u, i'd suggest to temporarily remove the prop (this prop is not mandatory, AdSlot will create valid slot ids for you under the hood when not passing them) until we can push a fix.

Thanks you!

**Comment by TomiTakussaari** [comment]
Yeah that would explain this, as I was reusing same slotIds for same type of page.

And yes, it does seem to work without slotIds, and as predefined slotIds are not a requirement for us (atleast currently), that works for us.
Thanks!

Should we close the ticket for now?

**Comment by jaanauati** [comment]
Great, let’s keep this ticket opened, I’ll review and push a fix in these coming days.

Thank you.

**Comment by tikkichan4** [comment]
Just faced this problem on 0.12.0. Is there any workaround for this? Thanks a lot !!!

**Comment by jaanauati** [comment]
@tikkichan4 couple of questions:
 - in your case, this error popped after updating to 0.12?
 - could u please share a code example?

thanks

**Comment by tikkichan4** [comment]
> @tikkichan4 couple of questions:
> 
> * in your case, this error popped after updating to 0.12?
> * could u please share a code example?
> 
> thanks
@jaanauati Sorry for late reply.
1) Tested on 0.11.2 and no error occurred. It probably only happens on 0.12
2) Tried to create a demo. Those codes is extracted from my project.
[](https://codesandbox.io/s/w234t)
But i faced another problem on this demo which is not happened on my project...
When i try to click the close button, the component does not re-render the nodes, it can be solved when downgrade to 0.11.2 or remove DFPSlotsProvider.

Hope this help! Thank you

**Comment by jaanauati** [comment]
@tikkichan4  I just published v0.12.2 to npm, this version includes fixes for the errors you reported, please give it a try and let me know if everything is ok at your end. 
https://codesandbox.io/s/new-l0soq

Thanks.

**Comment by tikkichan4** [comment]
> @tikkichan4 I just published v0.12.2 to npm, this version includes fixes for the errors you reported, please give it a try and let me know if everything is ok at your end.
> https://codesandbox.io/s/new-l0soq
> 
> Thanks.
@jaanauati Updated to 0.12.2 and worked again. Thank you for your fixes :D
